C51-122: Styling of Case view People tab

swastikpareek opened the pull request:
## Overview
This PR covers styling and code refactoring for Case View People's tab

## Before

<img width="1680" alt="c51-122- before" src="https://user-images.githubusercontent.com/3340537/46005325-a218ce80-c0d2-11e8-9ff3-2da159d5cd97.png">

## After
<img width="1680" alt="c51-122- after" src="https://user-images.githubusercontent.com/3340537/46005353-b1981780-c0d2-11e8-850d-ff33f1fbbf75.png">

## Technical Details
* SCSS vars 
 `$box-shadow-pane` : Pane box shadow
* Add Loading placeholder HTML and css
* Code refactor
* BEM classes 
  * `civicase__add-btn` : Btn styles with + icon on left
  * `civicase__people-tab__table-checkbox`: Specific styles for the checkbox
  * `civicase__people-tab-link`: for Tab link 
  * `civicase__people-tab__sub-tab`: Sub tab
  * `civicase__people-tab__search`: Search section
  * `civicase__people-tab__selection`: Selection pane
  * `civicase__people-tab__select-box`: Select box styling
  * `civicase__people-tab__filter`: Filter pane styling
  * `civicase__people-tab__filter-role`: Filter role dropdown styling
  * `civicase__people-tab__filter-alpha-pager`: Alphabet pagers
  * `civicase__people-tab__table-assign-icon`: Assign icon styles
  * `civicase__people-tab-counter`: Tab Counter styles
  * `civicase__people-tab__table-header`: People Table header
  * `civicase__people-tab__table-body`: People Table Body
  * `civicase__people-tab__table-column`: People Table column
  * `civicase__people-tab__table-column--first`: People Table first column
  * `civicase__people-tab__table-column--last`: People Table last column
* Update relationshipAPI return params to get the `start_date`

reneolivo commented on ang/civicase/CaseDetails--tabs--people.html:
```diff
@@ -0,0 +1,284 @@
+            <label for="select-role-{{ $index }}">
+              <div><strong>{{ role.role }}</strong></div>
+              <div><em>{{ role.description }}</em></div>
+              <div>{{ role.desc }}</div>
```
what is the difference between `role.description` and `role.desc` ?

swastikpareek replied:
Haven't written this. I think the role description is the description of the role and role.desc is the extra description that user fills when they update the assignee.

reneolivo commented on ang/civicase/CaseDetails--tabs--people.html:
```diff
@@ -0,0 +1,284 @@
+          </td>
+        </tr>
+        <tr ng-if="!roles.length">
+          <td colspan="9" style="text-align: center">
```
I think there's a `text-center` class we can use.

reneolivo commented on ang/civicase/CaseDetails--tabs--people.html:
```diff
@@ -0,0 +1,284 @@
+    </div>
+    <div class="civicase__people-tab__filter">
+      <paging
+      class="center-block paging-top"
```
Can we add one spacing for these attributes?

```html
<paging
  class="..."
  page="..."
  ...>
</paging>
```

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -1,14 +1,34 @@
+    var item = $scope.item;
+    var clients = _.indexBy(item.client, 'contact_id');
+    var relTypes = CRM.civicase.relationshipTypes;
+    var relTypesByName = _.indexBy(relTypes, 'name_b_a');
```
please sort them in alphabetical order.

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
-    }
     $scope.allRoles = [];
+    $scope.formatDate = DateHelper.formatDate;
+    $scope.getCaseRoles = getCaseRoles;
```
add a space between properties and method definitions.

reneolivo replied:
I think these should be defined with their functions to keep consistency:

```js
$scope. getCaseRoles = function () {
  // ...
};
```

Or is there something I'm missing?

swastikpareek replied:
`getCaseRoles ` is being used at other places in the controller. If define this function directly to the `$scope.getCaseRoles` then we have to order the code in such a way that other function can use these variables only after they are defined. So have followed this approach. Please feel free to ping me in case of a confusion

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
 
-    $scope.setSelectionMode = function(mode, tab) {
-      $scope[tab + 'SelectionMode'] = mode;
+    $scope.setTab = function (tab) {
```
Can we change this function?

```js
$scope.selectTab = function (tab) {
  $scope.selectedTab = tab;
};
```

reneolivo replied:
Missing docs

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
+      $scope.tab = tab;
+    };
+
+    $scope.setLetterFilter = function (letter, tab) {
```
missing docs

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
+      } else {
+        $scope[tab + 'AlphaFilter'] = letter;
+      }
+      if (tab === 'roles') {
```
space between the `if`s

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
+        $scope[tab + 'AlphaFilter'] = letter;
+      }
+      if (tab === 'roles') {
+        $scope.getCaseRoles();
```
If we go with the `$scope.tabs = {}` you can have something like this instead:

```js
$scope.tabs = {
  roles: { get: getCaseRoles },
  relations: { get: getRelations },
};

$scope.setLetterFilter = function (letter, tab) {
  // ...

  $scope.tabs[tab].get();
};
```

Also `get`, `getRelations`, and `getCaseRoles` are not the proper names for these functions. When a function starts with `get` there's an expectation that it will return a value, and this is not the case. It would be good to rename them to `loadRelations`, etc. but only if you have time and think this is feasible.

swastikpareek replied:
As per plan. We won't be fixing (refactoring) the logic of the already present functions if they work fine. As this will take too much time and disturb the sprint plan. However it will be a good add on to have documentation for these function.

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -203,28 +191,74 @@
 
-    $scope.$watch('item', getCaseRoles, true);
-    $scope.$watch('rolesFilter', getCaseRoles);
+    function formatRole (role) {
```
missing docs

reneolivo commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -203,28 +191,74 @@
-    $scope.setTab = function(tab) {
-      $scope.tab = tab;
-    };
+    function getCaseRoles () {
```
missing docs

reneolivo commented on scss/components/_buttons.scss:
```diff
@@ -0,0 +1,19 @@
+    content: '+';
+    display: inline-block;
+    font-size: 11px;
+    font-size: 13px;
```
`font-size` is repeated twice. Please make sure you are using the right size.

reneolivo commented on scss/components/_buttons.scss:
```diff
@@ -0,0 +1,19 @@
+    background: $crm-white;
+    border-radius: 100%;
+    color: $brand-primary;
+    content: '+';
```
why are we using `+` instead of the font awesome or material icon equivalent?

swastikpareek replied:
Couldn't find the icon in the material icon and font awesome library. If you can direct me to the one I can use it. However I always prefer doing icons with the css. They are cleaner, easily customizable and fast (performant).

reneolivo commented:
@swastikpareek I also found a few issues with the implementation:

* The roles filter does not seem to work for me (or maybe I'm doing something wrong)
* The checkbox takes a long while to change values and some times it does not change values when clicking it. 

![record](https://user-images.githubusercontent.com/1642119/46011747-e2595c00-c094-11e8-9429-be8eb5be2a98.gif)

swastikpareek commented on ang/civicase/CaseDetails--tabs--people.html:
```diff
@@ -0,0 +1,284 @@
+      <tbody class="civicase__people-tab__table-body">
+        <tr ng-repeat="role in roles" ng-class="{unassigned: !role.contact_id}">
+          <td class="civicase__people-tab__table-column civicase__people-tab__table-column--first">
+            <span ng-if="role.contact_id" class="civicase__checkbox" ng-if="!$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed" role="checkbox" aria-checked="{{activity.is_completed}}" tabindex="0" aria-labelledby="Mark complete/incomplete">
```
The whole code is a mess. I think this is missed from my eyes. Let me fix this.

swastikpareek commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
-        $scope.rolesPage = 1;
-      }
-      $scope.roles = _.slice(caseRoles, (25 * ($scope.rolesPage - 1)), 25 * $scope.rolesPage);
+    $scope.setSelectionMode = function (mode, tab) {
```
This is something which is already written. As per our plan we are not touching the existing code now. We are just refactoring.

swastikpareek commented on ang/civicase/CaseDetailsPeople.js:
```diff
@@ -23,113 +43,85 @@
-      }
-      $scope.roles = _.slice(caseRoles, (25 * ($scope.rolesPage - 1)), 25 * $scope.rolesPage);
+    $scope.setSelectionMode = function (mode, tab) {
+      $scope[tab + 'SelectionMode'] = mode;
```
I agree to your point. But as per the plan we are not touching the existing code. We are just refactor it.

swastikpareek commented on ang/civicase/CaseDetails--tabs--people.html:
```diff
@@ -0,0 +1,284 @@
+  </ul>
+  <!-- End Nav Tabs -->
+  <!-- Case Role Tab -->
+  <div class="civicase__people-tab__sub-tab" ng-show="tab == 'roles'">
```
Moving them into seprate `ng-include` is playing roughly with the $scope and breaking things.
Please let me know if you want a call for this.

swastikpareek commented:
@reneolivo : On debugging I found out that this was an issue with the current app (old dcase). However I have fixed the issue. Also the checkbox should work now.

As per the refactor plan we are not updating the existing logic if it is working fine. (As it will eat up time and this is what unfortunately, we don't have now). I have added docs to some function. Please check
